ts/src/analysis/analysis.ts

```diff
@@ -157,7 +155,10 @@ export default class Analysis implements Analyzer {
         if (f.name && f.name != "") {
             description.name = f.name;
         }
-        if (this.isNative(f)) {
+        // Checks if the function is a native function or that it exists in the map already
+        // since the same function can be invoked multiple times throughout a program.
+        if (this.isNativeMap.get(f.name) || this.isNative(f)) {
+            this.isNativeMap.set(f.name, true);
```

Comment:
Hmm, this is to cache whether or not a function is native, right?

I'm not sure this works in general, since native functions can be redefined as regular JS code.

For example:
```js
// Custom logging function that pipes through STDERR
console.log = function(data) {
    console.err("FROM STDOUT: " + data.toString());
}
```

Unfortunately, this stuff is not only possible but commonplace in JS. In fact, even Augur does this. Augur uses *polyfills* (https://developer.mozilla.org/en-US/docs/Glossary/Polyfill) to get some test cases to run, by redefining native functions with JavaScript implementations. Here are our polyfills: https://github.com/nuprl/augur/blob/main/ts/src/native/polyfill.ts

--------------

That means this optimization could unfortunately cause problems, if a native function is redefined to a JavaScript function and its "native status" changes. If a function changes from native to JS, the map will try to treat it as a native function and then incorrectly interpret the following instructions.

If you wanted to cache a function's native status, perhaps you could try tying that cache to the *function* itself, as opposed to its name. The function name "console.log" might have its native status changed during program execution, but the actual *function* `console.log` is immutable and therefore permanently native. This should be possible to cache, but you'd have to be careful not to interfere with garbage collection of these functions: this can be done by using a `WeakMap` instead of a `Map`.

Reply:
That makes sense, I hadn't really thought about how native functions can be redefined. I'll change it to the caching the function itself along with the _weakMap_ and seeing if it has similar improvements.